app/lib/account/backend.dart

```diff
@@ -0,0 +1,180 @@
+// Copyright (c) 2019, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:gcloud/db.dart';
+import 'package:gcloud/service_scope.dart' as ss;
+import 'package:googleapis/oauth2/v2.dart' as oauth2_v2;
+import 'package:http/http.dart' as http;
+import 'package:logging/logging.dart';
+import 'package:pool/pool.dart';
+import 'package:retry/retry.dart';
+
+import 'models.dart';
+
+final _logger = new Logger('pub.account.backend');
+final _retry = RetryOptions();
+
+/// The pub client's OAuth2 identifier.
+final _pubAudience = '818368855108-8grd2eg9tj9f38os6f1urbcvsq399u8n.apps.'
+    'googleusercontent.com';
+
+/// Sets the account backend service.
+void registerAccountBackend(AccountBackend backend) =>
+    ss.register(#_accountBackend, backend);
+
+/// The active account backend service.
+AccountBackend get accountBackend =>
+    ss.lookup(#_accountBackend) as AccountBackend;
+
+/// Represents the backend for the account handling and authentication.
+class AccountBackend {
+  final _pool = Pool(1);
+  final DatastoreDB _db;
+  final _defaultAuthProvider =
+      GoogleOauth2AuthProvider('dartlang-pub-1', _pubAudience);
+
+  AccountBackend(this._db);
+
+  Future close() async {
+    await _defaultAuthProvider.close();
+  }
+
+  Future<User> lookupOrCreateUser(String accessToken) async {
+    final auth = await _defaultAuthProvider.tryAuthenticate(accessToken);
+    if (auth == null) {
+      return null;
+    }
+    final compositeId = auth.compositeId;
+    final userKey = _db.emptyKey.append(User, id: compositeId);
+
+    final accessTime = DateTime.now().toUtc();
+    final user = await _retry.retry(() async {
+      User user = (await _db.lookup<User>([userKey])).single;
+      if (user == null) {
+        // create user
+        user = User()
+          ..parentKey = _db.emptyKey
+          ..id = compositeId
+          ..authProvider = auth.provider
+          ..authUserId = auth.userId
+          ..email = auth.email
+          ..emails = <String>[auth.email]
+          ..created = accessTime
+          ..updated = accessTime;
+        _db.commit(inserts: [user]);
+      } else {
+        // update user if e-mail has been changed
+        if (user.updateEmail(auth.email)) {
```

Comment:
having the updateEmail both mutate and return `changed` is a bit confusing to me.
Would it not be cleaner to write:
```
  if (user.email != auth.email) {...}
```
You are anyway repeating the updateEmail call below for the looked up user.